bids/modeling/statsmodels.py

```diff
@@ -0,0 +1,629 @@
+            List of strings indicating which entities to group on.
+
+        Returns
+        -------
+        A dictionary, where the keys are tuples of tuples and the values are lists
+        containing subsets of the original object list. Each outer tuple of the keys
+        represents a particular entity/value combination; each inner tuple has two
+        elements, where the first is the name of the entity and the second is the value.
+
+        Notes
+        -----
+        * The passed objects can be of any type, but alll elements in the list must
+        have an implemented .entities dictionary.
+        * Objects that do not have defined values for all of the requested
+        `entities` will be silently ignored.
+
+        Examples
+        --------
+        >>> bidsfile = namedtuple('bidsfile', ('entities',))
+        >>> bidsfile1 = bidsfile(entities={'subject': '01'})
+        >>> bidsfile2 = bidsfile(entities={'subject': '02'})
+        >>> groups = {(('subject', '01'),): [bidsfile1],
+        ...           (('subject', '02'),): [bidsfile2]}
+        >>> BIDSStatsModelsNode._build_groups(
+        ...     [bidsfile1, bidsfile2],
+        ...     ['subject']) == groups
+        True
+        """
+        if not group_by:
+            return {(): objects}
+
+        groups = defaultdict(list)
+
+        # sanitize grouping entities, otherwise weird things can happen
+        group_by = list(set(group_by) & VALID_GROUPING_ENTITIES)
+
+        # Get unique values in each grouping variable
+        entities = [obj.entities for obj in objects]
+        df = pd.DataFrame.from_records(entities).loc[:, group_by]
+        unique_vals = {col: df[col].dropna().unique().tolist() for col in group_by}
+
+        # Note: we can't just naively bucket objects based on the values of the
+        # grouping entities, because an object may have undefined values for
+        # one or more grouping variables. So we "fill in" the missing values by
+        # Looping over objects. For each one, identify all grouping variables
+        # with missing values. Loop over the cartesian product of the unique
+        # values of the missing entities, combining the permutation values with
+        # the base (present) grouping entities. Sort on this to produce the
+        # group key.
+        for i, row in df.iterrows():
+
+            defined = df.columns[df.iloc[i].notnull()].tolist()
+            base_ents = [(k, row[k]) for k in defined]
+            missing = df.columns[df.iloc[i].isnull()].tolist()
+
+            records = []
+
+            if missing:
+                product = itertools.product(*[unique_vals[col] for col in missing])
+                for perm in product:
+                    fill_ents = [(k, v) for (k, v) in dict(zip(missing, perm)).items()]
+                    records.append(base_ents + fill_ents)
+            else:
+                records.append(base_ents)
+
+            for rec in records:
+                grp_key = tuple(rec)
```

Comment:
I think you still want this sorted to avoid permutations?

```suggestion
                grp_key = tuple(sorted(rec))
```